### ISSUE TITLE
How to modify minimal server example to only reply to originating client ?

### ISSUE THREAD

**Issue by lifecycletest** [warmcat/libwebsockets#1796]
Hi,

  I new to WebSockets and I downloaded and tried out your library.  I was able to modify the "minimal-ws-server" example to work for me, but I will have multiple clients connecting to my server and sending commands, so I decided to switch to the "minimal-ws-server-ring" example because the client PCs can send commands concurrently to the server PC.

  I got that working too, but my problem is when the server's callback processes a client's request in the "LWS_CALLBACK_RECEIVE" case, it broadcasts the reply to all the clients (as documented).

  My question is seeing that I am a noob, what is the easiest way to fix this, so the server can queue multiple requests from different clients and each request's result is sent only to the originating client?

 _Note, the server processes the request by communicating with a slow I2C device, so a multi-threaded design would not increase its throughput and I would like to avoid this extra complexity if possible._

Thanks,
Phil

**Comment by lws-team** [comment]
Nobody here is going to start telling you to use threads randomly... lws is a singlethreaded event loop design.  But in i2c case, if you block the event loop while the transaction completes, depending on length of i2c transaction and bus speed, performance may become very slow if multiple clients.  If so you may want a thread that does the i2c part separate from lws event loop.

The ws-server-ring example is demoing how lws_ring supports a "shared world" where many clients can be shown the same consistent set of messages even when the clients catch up at different speeds.  That's very cool if that's what your application needs.

If each client lives in his own world, how to do it depends on details of what the server actually is for.  IIUI in your case clients want to take turns privately querying this shared i2c thing, and only see their own response later asynchronously... later because it's queued, and because i2c is too slow.  IIUI each individual client can only have one outstanding request at a time,

If that's it, ringbuffer is not the right tool.  The main difficulty is that clients may close and their wsi be destroyed at any time around the event loop, if you do things like store wsi pointer in an object buried in the ringbuffer, it becomes difficult to clean out references to a closing wsi from a ringbuffer.

Instead what I'd do for this is forget the ringbuffer, and use a doubly-linked list... add an lws_dll2_owner_t in the vhd and an lws_dll2_t in each pss.  When a new request comes in store the details in the pss and add the pss list object to the head of the list owner.  If you see a CLOSE meaning your client died, remove its list object from the list owner then, and if the closing guy was being worked on by the i2c thread, mark it as to be thrown away.  The i2c thread fulfils oldest requests first from the tail of the list owner, puts the result in the pss of the specific client and then calls lws_cancel_service() to alert the lws service thread to call lws_callback_on_writable() on the wsi the completed pss belongs to.

**Comment by lifecycletest** [comment]
Hi Andy,

  I switched back to the minimal-ws-server's callback function and spent a few more hours drilling through your code until I understood how it works at a very "high" level LoL.  Then I guessed that all I needed to do was to replace this code:
```
		/*
		 * let everybody know we want to write something on them
		 * as soon as they are ready
		 */
		lws_start_foreach_llp(struct per_session_data__minimal **,
				      ppss, vhd->pss_list) {
			lws_callback_on_writable((*ppss)->wsi);
		} lws_end_foreach_llp(ppss, pss_list);
```
with a single line:

>  lws_callback_on_writable(wsi)  // Reply only to "this" client

I tested it with 3 browser windows and it worked, and I also added a 5 second delay inside the I2C code and everything still worked and the server did not drop any requests, so I think I am GTG because the Linux networking stack should not drop any bytes if the server thread is busy for a few milliseconds doing other things.

I do appreciate your suggestion and I actually am starting to understand what you said even though I still don't know what "IIUI" means (google says it's a college in Islamabad LoL), but I fear I would screw  up the implementation and seeing as the client's will have a _synchronous_ API they are going to block waiting for the server, so the bandwidth will be set by the server and it's TCP port should never overflow.

Thanks,
Phil
